crates/ink/codegen/src/generator/metadata.rs

```diff
@@ -343,20 +343,28 @@ impl Metadata<'_> {
         let constructor_info = quote_spanned!(span =>
             < #storage_ident as ::ink::reflect::DispatchableConstructorInfo<#selector_id>>
         );
+
         quote_spanned!(span=>
             ::ink::metadata::ReturnTypeSpec::new(
-                if #constructor_info ::IS_RESULT {
+                if #constructor_info::IS_RESULT {
                     ::core::option::Option::Some(::ink::metadata::TypeSpec::with_name_str::<
-                        ::core::result::Result<
+                        ::ink::ConstructorResult<::core::result::Result<
                             (),
-                            #constructor_info ::Error
-                        >
+                            #constructor_info::Error
+                        >>
                     >(
-                        "core::result::Result"
+                        "ink_primitives::ConstructorResult"
                     )
                 )
                 } else {
-                    ::core::option::Option::None
+                    ::core::option::Option::Some(::ink::metadata::TypeSpec::with_name_str::<
```

Comment:
`ReturnTypeSpec` should now always expect a `TypeSpec` instead of an `Option<TypeSpec>`. There is one other usage to take care of for trait output types.

Reply:
I don't know enough about all the metadata types to be able to remove this and still get things compiling, so in the interest of time I'm gonna leave it a a follow-up

Reply:
Opened #1515